examples/markov_music/sketch.js

```diff
@@ -0,0 +1,253 @@
+    return false;
+  }
+};
+Node.prototype.applyForce = function(force) {
+  var f = p5.Vector.div(force,this.diameter);
+  this.acceleration.add(f);
+};
+Node.prototype.update = function() {
+  // Velocity changes according to acceleration
+  this.velocity.add(this.acceleration);
+  // position changes by velocity
+  this.position.add(this.velocity);
+};
+Node.prototype.display = function() {
+  noStroke();
+  var color = [200, 200, 200];
+  if (this.id == latestNodeId) {
+    // Highlight latest node
+    color = this.color;
+  }
+  // Fill circle if note-on, stroke circle if note-off
+  if (this.type == 1) {
+    noStroke();
+    fill(color[0], color[1], color[2]);
+  } else {
+    noFill();
+    strokeWeight(2);
+    stroke(color[0], color[1], color[2]);
+  }
+  ellipse(this.position.x,this.position.y,this.diameter,this.diameter);
+};
+Node.prototype.bounceOnBoundaries = function() {
+  if (this.position.x >= width || this.position.x <= 0) {
+    this.velocity.x *= -1;
+    this.position.x = constrain(this.position.x, 0, width);
+  }
+  if (this.position.y >= height || this.position.y <= 0) {
+    this.velocity.y *= -1;
+    this.position.y = constrain(this.position.y, 0, height);
+  }
+};
+
+// Graph data structure code adapted from 
+// http://blog.benoitvallon.com/data-structures-in-javascript/the-graph-data-structure/
+function Graph() {
+  this.nodes = [];
+  this.nodeIds = [];
+  this.edges = [];
+  this.numberOfEdges = 0;
+}
+Graph.prototype.findNode = function(node) {
+  for (var i=0; i<this.nodes.length; i++) {
+    if (node.isSimilar(this.nodes[i])) {
+      return i;
+    }
+  }
+  return -1; // Not found
+};
+Graph.prototype.registerNewNode = function(type, midiNoteNumber, duration) {
+  var node = new Node(0, type, midiNoteNumber, duration);
+  var nodeId = graph.findNode(node);
+  if (nodeId == -1) { // If necessary, create the node
+    nodeId = this.nodes.length;
+    this.addNode(node);
+  }
+  node.id = nodeId;
+  if (latestNodeId != null) { // On initialization it will be null
+    // Add an edge from the previous node to this one
+    this.addEdge(latestNodeId, nodeId);
+  }
+  // Update the latest node ID
+  latestNodeId = nodeId;
+};
+Graph.prototype.addNode = function(node) {
+  var nodeId = this.nodes.length;
+  this.nodeIds.push(nodeId);
+  this.nodes.push(node);
+  this.edges[nodeId] = [];
+};
+Graph.prototype.addEdge = function(nodeId1, nodeId2) {
+  this.edges[nodeId1].push(nodeId2);
+  this.numberOfEdges++;
+};
+Graph.prototype.drawEdges = function() {
+  // Draw all edges leading away from this node
+  strokeWeight(1);
+  for (var i=0; i<graph.edges.length; i++) {
+    var startNode = i;
+    if (startNode == latestNodeId) { // Highlight the latest node's edges
+      stroke(graph.nodes[startNode].color[0], graph.nodes[startNode].color[1], graph.nodes[startNode].color[2], 100);
+    } else {
+      stroke(200, 100);
+    }
+    for (var j=0; j<graph.edges[i].length; j++) {
+      var endNode = graph.edges[i][j];
```

Comment:
this is a directed graph. Perhaps a little arrow will help illustrate the concept?